idle-crafting.lua

```diff
@@ -6,6 +6,55 @@ local widgets = require('gui.widgets')
 local repeatutil = require("repeat-util")
 local orders = require('plugins.orders')
 
+---iterate over input materials of workshop with stockpile links
+---@param workshop df.building_workshopst
+---@param action fun(item:df.item):any
+local function for_inputs(workshop, action)
+    if #workshop.profile.links.take_from_pile == 0 then
+        dfhack.error('workshop has no links')
+    else
```

Comment:
you can turn the `else` into an `end` and save some indentation in the rest of the function. `dfhack.error` throws and does not allow the function to continue.